### PR TITLE
[Linux][Windows] Modify CPU and memory hotadd test case

### LINKED ISSUE #6: [Bug] [Windows][Intermittent] Failed to extract VMware Tools ISO to get PVSCSI driver for Windows installation
##### SUMMARY

When excuting Windows autoinstall on PVSCSI disk, it failed to extract ISO file to get PVSCSI driver and reported 7z is missed. But when I run it at the second time, the ISO can be extracted successfully.
```
2021-04-02 15:32:44,002 | TASK [Extract ISO file to get PVSCSI driver] ***************
task path: /root/ansible-vsphere-gos-validation/windows/deploy_vm/get_pvscsi_driver.yml:41
fatal: [localhost]: FAILED! => {
    "changed": false,
    "cmd": "mount -o loop,ro \"/root/ansible-vsphere-gos-validation/cache/VMware-tools-windows-11.2.5-17337674.iso\" \"/tmp/tmpetj7zmuc\"",
    "dest": "/root/ansible-vsphere-gos-validation/cache/",
    "files": [],
    "gid": 0,
    "group": "root",
    "image": "/root/ansible-vsphere-gos-validation/cache/VMware-tools-windows-11.2.5-17337674.iso",
    "invocation": {
        "module_args": {
            "dest": "/root/ansible-vsphere-gos-validation/cache/",
            "executable": null,
            "files": [
                "Program Files/VMware/VMware Tools/Drivers/pvscsi/Win8/amd64/pvscsi.cat",
                "Program Files/VMware/VMware Tools/Drivers/pvscsi/Win8/amd64/pvscsi.inf",
                "Program Files/VMware/VMware Tools/Drivers/pvscsi/Win8/amd64/pvscsi.sys",
                "Program Files/VMware/VMware Tools/Drivers/pvscsi/Win8/amd64/txtsetup.oem"
            ],
            "force": true,
            "image": "/root/ansible-vsphere-gos-validation/cache/VMware-tools-windows-11.2.5-17337674.iso"
        }
    },
    "mode": "0777",
    "msg": "Failed to mount ISO image '/root/ansible-vsphere-gos-validation/cache/VMware-tools-windows-11.2.5-17337674.iso' to '/tmp/tmpetj7zmuc', and we could not find executable '7z'.",
    "owner": "root",
    "rc": 32,
    "size": 135,
    "state": "directory",
    "stderr": "mount: /tmp/tmpetj7zmuc: failed to setup loop device for /root/ansible-vsphere-gos-validation/cache/VMware-tools-windows-11.2.5-17337674.iso.\n",
    "stderr_lines": [
        "mount: /tmp/tmpetj7zmuc: failed to setup loop device for /root/ansible-vsphere-gos-validation/cache/VMware-tools-windows-11.2.5-17337674.iso."
    ],
    "stdout": "",
    "stdout_lines": [],
    "uid": 0
}
```

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION

```paste below
ansible 2.10.7
  config file = /root/ansible-vsphere-gos-validation/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.9.1 (default, Feb 24 2021, 22:17:18) [GCC 10.2.0]
```

##### CONFIGURATION

```paste below
ANSIBLE_SSH_RETRIES(/root/ansible-vsphere-gos-validation/ansible.cfg) = 5
DEFAULT_CALLBACK_PLUGIN_PATH(/root/ansible-vsphere-gos-validation/ansible.cfg) = ['/root/ansible-vsphere-gos-validation/plugin']
DEFAULT_CALLBACK_WHITELIST(/root/ansible-vsphere-gos-validation/ansible.cfg) = ['timer']
DISPLAY_SKIPPED_HOSTS(/root/ansible-vsphere-gos-validation/ansible.cfg) = False
RETRY_FILES_ENABLED(/root/ansible-vsphere-gos-validation/ansible.cfg) = False
```

##### ENVIRONMENT / OS

```
Testbed version: vSphere 7.0 U2
OS: Windows Server 2022
new_vm: true
vm_deploy_method: iso
secureboot_enabled: false
firmware: efi
boot_disk_controller: paravirtual
network_adapter_type: vmxnet3
cdrom_controller_type: ide
```

##### STEPS TO REPRODUCE

Deploy a new VM and set boot disk controller to paravirtual.

##### EXPECTED RESULTS

Windows can be installed successfully.

##### ACTUAL RESULTS

Windows installation failed at extracting VMware Tools ISO for getting pvscsi driver.

### LINKED ISSUE #12: [Bug] [Windows] memory_hot_add_basic failed due to cpu limitation
##### SUMMARY

Windows memory_hot_add_basic failed due to cpu is not enough.
```

2021-04-07 07:39:23,007 | Failed at Play [memory_hot_add_basic] **********************

2021-04-07 07:39:23,007 | TASK [Set VM power state to 'powered-on'] ******************
task path: /root/ansible-vsphere-gos-validation/common/vm_set_power_state.yml:34
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "pyVmomi.VmomiSupport.NotEnoughCpus: (vim.fault.NotEnoughCpus) {\n   dynamicType = <unset>,\n   dynamicProperty = (vmodl.DynamicProperty) [],\n   msg = 'Virtual machine requires 8 CPUs to operate, but the host hardware only provides 4.',\n   faultCause = <unset>,\n   faultMessage = (vmodl.LocalizableMessage) [],\n   numCpuDest = 4,\n   numCpuVm = 8\n}\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/tmp/ansible-tmp-1617781160.9507868-13324-39469946055191/AnsiballZ_vmware_guest_powerstate.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/tmp/ansible-tmp-1617781160.9507868-13324-39469946055191/AnsiballZ_vmware_guest_powerstate.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/tmp/ansible-tmp-1617781160.9507868-13324-39469946055191/AnsiballZ_vmware_guest_powerstate.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.community.vmware.plugins.modules.vmware_guest_powerstate', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python3.9/runpy.py\", line 210, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.9/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_vmware_guest_powerstate_payload_t6k9rlw8/ansible_vmware_guest_powerstate_payload.zip/ansible_collections/community/vmware/plugins/modules/vmware_guest_powerstate.py\", line 273, in <module>\n  File \"/tmp/ansible_vmware_guest_powerstate_payload_t6k9rlw8/ansible_vmware_guest_powerstate_payload.zip/ansible_collections/community/vmware/plugins/modules/vmware_guest_powerstate.py\", line 261, in main\n  File \"/tmp/ansible_vmware_guest_powerstate_payload_t6k9rlw8/ansible_vmware_guest_powerstate_payload.zip/ansible_collections/community/vmware/plugins/module_utils/vmware.py\", line 843, in set_vm_power_state\n  File \"/tmp/ansible_vmware_guest_powerstate_payload_t6k9rlw8/ansible_vmware_guest_powerstate_payload.zip/ansible_collections/community/vmware/plugins/module_utils/vmware.py\", line 83, in wait_for_task\n  File \"<string>\", line 3, in raise_from\nansible_collections.community.vmware.plugins.module_utils.vmware.TaskError: ('Virtual machine requires 8 CPUs to operate, but the host hardware only provides 4.', None)\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

```

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION

```paste below



```

##### CONFIGURATION

```paste below



```

### LINKED ISSUE #15: [Bug][Linux][Windows] test case cpu_hot_add_basic/memory_hot_add_basic/cpu_multicores_per_socket improvement
##### SUMMARY

The logic of generating CPU/memory hotadd increase value list can be same and reused in test cases, cpu_hot_add_basic/memory_hot_add_basic/cpu_multicores_per_socket for Linux and Windows.

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION

```paste below
ansible 2.10.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.5 (default, Apr  1 2018, 05:46:30) [GCC 7.3.0]
```

##### CONFIGURATION

```paste below



```

##### ENVIRONMENT / OS

Linux and Windows

##### EXPECTED RESULTS

Test cases and tasks improvement

##### ACTUAL RESULTS

There are some tasks can be reused among test cases.

### PR DESCRIPTION
**Pull request by Tomorrow9**
Signed-off-by: dianew <dianew@vmware.com>

Fixes #6
Fixes #12 
Fixes #15 
Modify CPU and memory hotadd test cases for Linux and Windows to use the same task to generate the hotadd value list according to the actual available ESXi host resource and the configured max test value.
Affected test cases:
```
cpu_hot_add_basic
memory_hot_add_basic
```

### CODE REVIEW

File: linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2021 VMware, Inc.
```
Comment (keirazhang) [inline linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml:1]: I think we can move this task file to common folder as it is been used by both Linux and Windows.

File: linux/gosv_testcase_list.yml [hunk 2]
```diff
@@ -1,29 +1,29 @@
...
+#- import_playbook: network_device_ops/vmxnet3_network_device_ops.yml
+#- import_playbook: cpu_multicores_per_socket/cpu_multicores_per_socket.yml
+#- import_playbook: guest_customization/gosc_perl_dhcp.yml
+#- import_playbook: guest_customization/gosc_perl_staticip.yml
+#- import_playbook: guest_customization/gosc_cloudinit_dhcp.yml
+#- import_playbook: guest_customization/gosc_cloudinit_staticip.yml
+#- import_playbook: vhba_hot_add_remove/paravirtual_vhba_device_ops.yml
+#- import_playbook: vhba_hot_add_remove/lsilogic_vhba_device_ops.yml
+#- import_playbook: vhba_hot_add_remove/lsilogicsas_vhba_device_ops.yml
+#- import_playbook: open_vm_tools/ovt_verify_uninstall.yml
```
Comment (keirazhang) [inline linux/gosv_testcase_list.yml:29]: Cases should not be commented.

**Review by keirazhang — DISMISSED** [review]
LGTM. Thanks!

**Comment by Tomorrow9** [comment]
Testing passed on CentOS 8.3:
```
TASK [debug] *******************************************************************
task path: /root/ansible-vsphere-gos-validation/common/esxi_get_cpu_mem_info.yml:21
ok: [localhost] => {
    "msg": [
        "ESXi host logical processor: 40",
        "ESXi host total memory in MB: 104738.28125",
        "ESXi host consumed memroy in MB: 23167",
        "ESXi host free memory in MB: 81571"
    ]
}
Read vars_file '{{ testing_vars_file | default('../../vars/test.yml') }}'

TASK [Set fact of the maximum vCPU number of hotadd test] **********************
task path: /root/ansible-vsphere-gos-validation/linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml:15
ok: [localhost] => {
    "ansible_facts": {
        "max_num_cpus": "16"
    },
    "changed": false
}
Read vars_file '{{ testing_vars_file | default('../../vars/test.yml') }}'

TASK [debug] *******************************************************************
task path: /root/ansible-vsphere-gos-validation/linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml:18
ok: [localhost] => {
    "msg": "Maximum VM vCPU number of CPU hotadd test: 16"
}
Read vars_file '{{ testing_vars_file | default('../../vars/test.yml') }}'

TASK [Set fact of the vCPU number increase list] *******************************
task path: /root/ansible-vsphere-gos-validation/linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml:26
ok: [localhost] => {
    "ansible_facts": {
        "cpu_hotadd_num_list": [
            1,
            2,
            3,
            4
        ]
    },
    "changed": false
}
...
TASK [Display the test results] ************************************************
task path: /root/ansible-vsphere-gos-validation/linux/cpu_hot_add_basic/cpu_hot_add_basic.yml:39
ok: [localhost] => {
    "cpu_hotadd_results": [
        "vCPU hotadd succeeds: 2->3",
        "vCPU hotadd succeeds: 3->5",
        "vCPU hotadd succeeds: 5->8",
        "vCPU hotadd succeeds: 8->12"
    ]
}
```

**Comment by Tomorrow9** [comment]
```
TASK [Display the test results] ************************************************
task path: /root/ansible-vsphere-gos-validation/linux/memory_hot_add_basic/memory_hot_add_basic.yml:35
ok: [localhost] => {
    "mem_hotadd_results": [
        "Memory hotadd succeeds: 2048->3072",
        "Memory hotadd succeeds: 3072->5120",
        "Memory hotadd succeeds: 5120->8192",
        "Memory hotadd succeeds: 8192->12288"
    ]
}
```
```
TASK [Display the test results] ************************************************
task path: /root/ansible-vsphere-gos-validation/linux/cpu_multicores_per_socket/cpu_multicores_per_socket.yml:39
ok: [localhost] => {
    "cpu_multicores_per_socket_results": [
        "VM vCPU cores per socket set succeeds: 3",
        "VM vCPU cores per socket set succeeds: 5",
        "VM vCPU cores per socket set succeeds: 8",
        "VM vCPU cores per socket set succeeds: 12"
    ]
}
```

File: linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml (hunk 1 above)
Comment (Tomorrow9) [reply linux/cpu_hot_add_basic/generate_cpu_hot_add_list.yml:1]: These common tasks are test case specific, so I think we can keep it there, if there is better way for common used tasks in test cases for both Linux and Windows, we can change it then. Thanks.

**Comment by Tomorrow9** [comment]
Testing passed on Windows 10 64bit:
```
TASK [Display the test results] ************************************************
task path: /root/ansible-vsphere-gos-validation/windows/memory_hot_add_basic/memory_hot_add_basic.yml:41
ok: [localhost] => {
    "mem_hotadd_results": [
        "Memory hotadd succeeds: 4096->5120 with vCPU number: 2",
        "Memory hotadd succeeds: 5120->7168 with vCPU number: 2",
        "Memory hotadd succeeds: 7168->10240 with vCPU number: 2",
        "Memory hotadd succeeds: 10240->14336 with vCPU number: 2",
        "Memory hotadd succeeds: 4096->5120 with vCPU number: 4",
        "Memory hotadd succeeds: 5120->7168 with vCPU number: 4",
        "Memory hotadd succeeds: 7168->10240 with vCPU number: 4",
        "Memory hotadd succeeds: 10240->14336 with vCPU number: 4",
        "Memory hotadd succeeds: 4096->5120 with vCPU number: 6",
        "Memory hotadd succeeds: 5120->7168 with vCPU number: 6",
        "Memory hotadd succeeds: 7168->10240 with vCPU number: 6",
        "Memory hotadd succeeds: 10240->14336 with vCPU number: 6",
        "Memory hotadd succeeds: 4096->5120 with vCPU number: 8",
        "Memory hotadd succeeds: 5120->7168 with vCPU number: 8",
        "Memory hotadd succeeds: 7168->10240 with vCPU number: 8",
        "Memory hotadd succeeds: 10240->14336 with vCPU number: 8"
    ]
}
```
```
TASK [Display the test results] ************************************************
task path: /root/ansible-vsphere-gos-validation/windows/cpu_hot_add_basic/cpu_hot_add_basic.yml:46
ok: [localhost] => {
    "cpu_hotadd_results": [
        "vCPU hotadd succeeds: 2->4"
    ]
}
```
```
TASK [Display the test results] ************************************************
task path: /root/ansible-vsphere-gos-validation/windows/cpu_multicores_per_socket/cpu_multicores_per_socket.yml:33
ok: [localhost] => {
    "cpu_multicores_per_socket_results": [
        "VM vCPU cores per socket set succeeds: 3",
        "VM vCPU cores per socket set succeeds: 5",
        "VM vCPU cores per socket set succeeds: 8",
        "VM vCPU cores per socket set succeeds: 12"
    ]
}
```

**Review by keirazhang — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**